File: velolabs_repos/lattis_android/app/src/main/java/com/lattis/ellipse/presentation/ui/base/fragment/bluetooth/BaseBluetoothFragment.java
package com.lattis.ellipse.presentation.ui.base.fragment.bluetooth;

import android.Manifest;
import android.app.Activity;
import android.bluetooth.BluetoothAdapter;
import android.content.Intent;
import android.content.IntentSender;
import android.content.pm.PackageManager;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;

import com.google.android.gms.common.api.ResolvableApiException;
import com.lattis.ellipse.domain.model.LocationSettingsResult;
import com.lattis.ellipse.presentation.ui.base.fragment.BaseFragment;
import com.lattis.ellipse.presentation.ui.utils.PopUpActivity;

import io.lattis.ellipse.R;
import permissions.dispatcher.NeedsPermission;
import permissions.dispatcher.RuntimePermissions;

import static android.app.Activity.RESULT_OK;
import static com.lattis.ellipse.presentation.ui.base.fragment.bluetooth.BaseBluetoothFragmentPermissionsDispatcher.getLocationPermissionWithPermissionCheck;

@RuntimePermissions
public abstract class BaseBluetoothFragment<Presenter extends BluetoothFragmentPresenter> extends BaseFragment<Presenter>
        implements BluetoothFragmentView {

    private static final int REQUEST_CODE_ENABLE_BLUETOOTH = 5647;
    private final int REQUEST_CHECK_LOCATION_SETTINGS = 2106;
    private final int REQUEST_CODE_SYSTEM_LOCATION_PERMISSION_DENIED = 2107;
    private final int REQUEST_CODE_LOCATION_PERMISSION_DENIED = 2108;


    @Override
    public void requestEnableBluetooth() {
        startActivityForResult(new Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE),
                REQUEST_CODE_ENABLE_BLUETOOTH);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if(requestCode == REQUEST_CODE_ENABLE_BLUETOOTH && resultCode == RESULT_OK){
            getPresenter().onBluetoothEnabled();
            onBluetoothEnabled();
        }else if(requestCode == REQUEST_CHECK_LOCATION_SETTINGS) {
            if (resultCode == RESULT_OK) {
                getPresenter().onLocationSettingsON();
                onLocationSettingsON();
            } else {
                PopUpActivity.launchForResultFromFragment(this,getActivity(), REQUEST_CODE_SYSTEM_LOCATION_PERMISSION_DENIED, getString(R.string.notice),
                        getString(R.string.privacy_location_alert_text), "", getString(R.string.privacy_location_submit_button));
            }
        }else if (requestCode==REQUEST_CODE_SYSTEM_LOCATION_PERMISSION_DENIED ){
            getPresenter().getLocationSetting();
        }else if(requestCode == REQUEST_CODE_LOCATION_PERMISSION_DENIED){
            getLocationPermissionWithPermissionCheck(this);
        }
    }

    @Override
    public void onBluetoothEnabled() {

    }


    protected void checkForLocation() {
        getLocationPermissionWithPermissionCheck(this);
    }

    @NeedsPermission(Manifest.permission.ACCESS_FINE_LOCATION)
    public void getLocationPermission() {
        getPresenter().getLocationSetting();
    }



    @Override
    public void onLocationSettingsPermissionRequired(LocationSettingsResult locationSettingsResult) {
        try {
            // Show the dialog by calling startResolutionForResult(),
            // and check the result in onActivityResult().
            ResolvableApiException resolvable = (ResolvableApiException) locationSettingsResult.getApiException();
            startIntentSenderForResult(resolvable.getResolution().getIntentSender(),REQUEST_CHECK_LOCATION_SETTINGS , null, 0, 0, 0, null);
        } catch (IntentSender.SendIntentException e) {
            // Ignore the error.
            onLocationSettingsNotAvailable();
        }
    }

    @Override
    public void onLocationSettingsON() {

    }

    @Override
    public void onLocationSettingsNotAvailable() {

    }


    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (ActivityCompat.checkSelfPermission(getActivity(), Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            PopUpActivity.launchForResultFromFragment(this,getActivity(), REQUEST_CODE_LOCATION_PERMISSION_DENIED, getString(R.string.notice),
                    getString(R.string.privacy_location_alert_text), "", getString(R.string.privacy_location_submit_button));
        }
        BaseBluetoothFragmentPermissionsDispatcher.onRequestPermissionsResult(this, requestCode, grantResults);
    }
}

File: velolabs_repos/lattis_android_v2/app/src/main/java/com/lattis/lattis/presentation/base/fragment/bluetooth/BaseBluetoothFragment.kt
package com.lattis.lattis.presentation.base.fragment.bluetooth

import android.Manifest
import android.app.Activity
import android.bluetooth.BluetoothAdapter
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.util.Log
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.app.ActivityCompat
import com.lattis.lattis.presentation.base.fragment.usercurrentlocation.BaseUserCurrentStatusFragment
import io.lattis.lattis.R
import io.reactivex.rxjava3.annotations.NonNull
import permissions.dispatcher.NeedsPermission


abstract class BaseBluetoothFragment<Presenter : BaseBluetoothFragmentPresenter<V>,V: BaseBluetoothFragmentView> :
    BaseUserCurrentStatusFragment<Presenter, V>(), BaseBluetoothFragmentView {


    val REQUEST_CODE_ENABLE_BLUETOOTH = 5647
    val REQUEST_CODE_BLUETOOTH_AFTER_DENY = 5648
    val REQUEST_CODE_BLUETOOTH_SCAN_CONNECT_AFTER_DENY = 5649


    private fun checkPermission(@NonNull permission: String): Boolean {
        return (ActivityCompat.checkSelfPermission(requireContext(), permission)
                === PackageManager.PERMISSION_GRANTED)
    }

    override fun checkBLEPermissions(){
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val deniedPermissions: ArrayList<String> = ArrayList()
            if(!checkPermission(Manifest.permission.BLUETOOTH_SCAN))
                deniedPermissions.add(Manifest.permission.BLUETOOTH_SCAN);
            if(!checkPermission(Manifest.permission.BLUETOOTH_CONNECT))
                deniedPermissions.add(Manifest.permission.BLUETOOTH_CONNECT);
            if(deniedPermissions.isEmpty()){
                presenter.startLockConnectionForBLE()
            }else{
                requestMultiplePermissions.launch(deniedPermissions.toTypedArray())
            }
        }else{
            presenter.startLockConnectionForBLE()
        }
    }

    override fun requestEnableBluetooth() {
        startActivityForResult(
            Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE),
            REQUEST_CODE_ENABLE_BLUETOOTH
        )
    }

    private val requestMultiplePermissions =
        registerForActivityResult(ActivityResultContracts.RequestMultiplePermissions()) { permissions ->
            var anyPermissionNotGranted = false
            permissions.entries.forEach {
                Log.d("test006", "${it.key} = ${it.value}")
                if(!it.value) {
                    anyPermissionNotGranted = true
                }
            }

            if(anyPermissionNotGranted){
                showBluetoothRequiredPop(REQUEST_CODE_BLUETOOTH_SCAN_CONNECT_AFTER_DENY)
            }else{
                presenter.startLockConnectionForBLE()
            }
        }

    override fun onBluetoothEnabled() {

    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == REQUEST_CODE_ENABLE_BLUETOOTH && resultCode == Activity.RESULT_OK) {
            presenter.onBluetoothEnabled()
            onBluetoothEnabled()
        }else if (requestCode == REQUEST_CODE_ENABLE_BLUETOOTH && resultCode == Activity.RESULT_CANCELED) {
            showBluetoothRequiredPop(REQUEST_CODE_BLUETOOTH_AFTER_DENY)
        }else if( requestCode == REQUEST_CODE_BLUETOOTH_AFTER_DENY){
            requestEnableBluetooth()
        }else if( requestCode == REQUEST_CODE_BLUETOOTH_SCAN_CONNECT_AFTER_DENY){
            checkBLEPermissions()
        }
    }

    fun showBluetoothRequiredPop(requestCode:Int){
        launchPopUpActivity(
            requestCode,
            getString(R.string.bluetooth_access_alert_message),
            null,
            null,
            getString(R.string.general_btn_ok),
            null,
            null,
            null
        )
    }
}